### ISSUE TITLE
Handle driver submitting a task with a function that hasn't been registered.

### ISSUE THREAD

**Issue by robertnishihara** [amplab/ray#105]
Consider the script
```python
import ray
import ray.services as services
import test_functions

# assume that we run this script from inside ray/test/
test_dir = os.path.dirname(os.path.abspath(__file__))
test_path = os.path.join(test_dir, "test_worker.py")
services.start_singlenode_cluster(return_drivers=False, num_workers_per_objstore=1, worker_path=test_path)

x = test_functions.keyword_fct1(1)
xval = ray.pull(x) # xval should be "1 hello"
```
Two things happen:

1. The line `servics.start_singlenode_cluster` starts a worker process, which connects to the scheduler and registers all of the functions in the `test_functions` module (that is, it lets the scheduler know that it knows how to execute those functions).
2. The line `x = test_functions.keyword_fct1(1)` submits a task to the scheduler for executing the function `test_functions.keyword_fct1`.

There are two possibilities.

- If 1) happens before 2), then everything works great.
- If 2) happens before 1), then things are trickier. The method `SubmitTask` currently throws a fatal error. The advantage of throwing a fatal error is that a lot of the time when 2) happens before 1) it will be because the user forgot to register a particular module, and we will want to draw this to the user's attention to make debugging easier. The disadvantage of throwing a fatal error is of course that this situation can happen legitimately. For example, if `test_worker.py` takes a while before registering its functions (e.g., it first creates a tensorflow model), then 2) will probably happen first. In this situation, what should we do? Perhaps the answer is that the call to `x = test_functions.keyword_fct1(1)` should block until the function has been registered on some worker. The advantages of this are
    - It should be easier to debug if the user forgot to import a module.
    - If we are waiting for a slow worker, then it won't fail.
    - The function `SubmitTask` needs to return the correct number of object references to the driver program. Right now, it gets that information when the function gets registered, so if we wait, then we will have that information. If we don't wait and want to return immediately, then that information will have to be in the `SubmitTaskRequest`, which feels a bit ugly.

Thoughts? @pcmoritz @mehrdadn @Wapaul1.

**Comment by mehrdadn** [comment]
Suggestion: Have the scheduler return an error code of some sort, then have the client retry. Hopefully do this lower in the client's stack so it can be done automatically by default and doesn't affect every single talk. This should let the clients customize their behavior from pure blocking (e.g. timeouts, exponential backoff, etc.).

**Comment by robertnishihara** [comment]
Addressed by  #108.
